lib/MusicBrainz/LWP/TimelyUserAgent.pm

```diff
@@ -0,0 +1,110 @@
+package MusicBrainz::LWP::TimelyUserAgent;
+
+use Moose;
+
+use DBDefs;
+use HTTP::Status qw( HTTP_SERVICE_UNAVAILABLE );
+use MusicBrainz::Server::Exceptions qw( GenericTimeout );
+use POSIX qw( SIGALRM );
+use Try::Tiny;
+
+extends 'LWP::UserAgent';
+
+use MooseX::Types::Moose qw( Int );
+
+has 'sending_timeout' => (
+    is => 'rw',
+    isa => 'Int',
+);
+
+sub new {
+    my ($class, %args) = @_;
+
+    $args{agent} = DBDefs->LWP_USER_AGENT unless exists $args{agent};
+
+    my $sending_timeout =
+        exists $args{sending_timeout}
+            ? delete $args{sending_timeout}
+            : 1;
+
+    my $obj = $class->SUPER::new(%args);
+
+    return $class->meta->new_object(
+        __INSTANCE__ => $obj,
+        sending_timeout => $sending_timeout,
+    );
+}
+
+around send_request => sub {
+    my ($orig, $self, @args) = @_;
+
+    my $sending_timeout = $self->sending_timeout // 0;
+    my $request_timeout = $self->timeout // 0;
+
+    return $self->$orig(@args) if $sending_timeout <= 0 || $request_timeout <= 0;
+
+    my ($request, $arg, $size) = @args;
+
+    my $action_timeout = $sending_timeout + $request_timeout;
+
+    alarm($action_timeout);
+
+    my $action = POSIX::SigAction->new(sub {
+        MusicBrainz::Server::Exceptions::GenericTimeout->throw(
+            $request->method . ' ' . $request->uri .
+            " took more than $action_timeout seconds (sending included)"
+        );
+    });
+    $action->safe(1);
+    POSIX::sigaction(SIGALRM, $action);
+
+    my $response;
+    try {
+        $response = $self->$orig(@args);
+    } catch {
+        my $err = $_;
+        if (ref($err) eq 'MusicBrainz::Server::Exceptions::GenericTimeout') {
```

Comment:
It catches the `GenericTimeout` but there is a bug while forging/retrieving the response.
See commit https://github.com/metabrainz/musicbrainz-server/pull/1102/commits/5d3dc58ef08c4625922f0e361fe8e51262ed65d3 and below log excerpt on `GET /search?query=MBS-10232&type=release`:

```
[debug] Sleeping until the alarm rings...
[debug] Throwing exception from signal
[debug] Catching error of type 'MusicBrainz::Server::Exceptions::GenericTimeout'...
[debug] Error message = 'GET http://search.musicbrainz.org/ws/2/release/?query=MBS-10232&offset=0&max=25&fmt=jsonnew&dismax=true&web=1 took more than 1 seconds (sending included)'
[debug] Forging HTTP::Response with code '503'...
[debug] Response code = 'HTTP::Request=HASH(0x5620f85638d0)'
[debug] Response message = '503'
[debug] Catching search error...
[debug] Search returned code = 'HTTP::Request=HASH(0x5620f85638d0)'
[debug] Search returned error = 'GET http://search.musicbrainz.org/ws/2/release/?query=MBS-10232&offset=0&max=25&fmt=jsonnew&dismax=true&web=1 took more than 1 seconds (sending included)'
```

Response code should be 503 and response message should be the same as error message.

What am I missing?

Reply:
Seems to be due to https://github.com/metabrainz/musicbrainz-server/pull/1102/files/efd59b627bd8ed562d42d7d95663d39b8a2c3fcd#diff-045d15826466d1e497f53b359fbf1139R67

`$self->SUPER::_new_response` implicitly passes `$self` as the first argument, which `_new_response` doesn't expect. It can be replaced with `LWP::UserAgent::_new_response(...)`.

Also looks like the `GenericTimeout` stack trace gets logged by plackup even though it's caught, which is odd and not ideal. :/